zenoh-flow-daemon/src/daemon.rs

```diff
@@ -106,6 +106,45 @@ impl TryFrom<DaemonConfig> for Daemon {
     fn try_from(config: DaemonConfig) -> std::result::Result<Self, Self::Error> {
         use zenoh::prelude::sync::*;
 
+        // Loading Zenoh configuration
+        let zconfig = get_zenoh_config(&config.zenoh_config)?;
+
+        // Creates the zenoh session.
+        let session = Arc::new(zenoh::open(zconfig).res()?);
+
+        Self::from_session_and_config(session, config)
+    }
+}
+
+impl Daemon {
+    /// Creates a new `Daemon` from the given parameters.
+    pub fn new(
+        z: Arc<zenoh::Session>,
+        ctx: RuntimeContext,
+        config: RuntimeConfig,
+        pool_size: usize,
+    ) -> Self {
+        let store = DataStore::new(z.clone());
+
+        let runtime = Runtime::new(z, ctx.clone(), config.clone());
+
+        let c_runtime = runtime.clone();
+        let new_worker = Arc::new(move |id, rx, hlc| {
+            Box::new(Worker::new(id, rx, hlc, c_runtime)) as Box<dyn WorkerTrait>
+        });
+
+        let mut workers =
+            WorkerPool::new(pool_size, store, config.uuid, ctx.hlc.clone(), new_worker);
+        workers.start();
+
+        Self {
+            runtime,
+            worker_pool: Arc::new(RwLock::new(workers)),
+            ctx,
+        }
+    }
+
+    pub fn from_session_and_config(z: Arc<zenoh::Session>, config: DaemonConfig) -> ZFResult<Self> {
```

Comment:
Question: why not have a single function to create a Daemon that takes as parameters a Zenoh session and a DaemonConfig?

We could provide a `impl Default for DaemonConfig` that fetches all default values. A builder pattern could even be added to help creating (+ validating) the configuration.

Did I miss something?

Reply:
It is done this way to avoid duplicating code between a plugin runtime and a standalone one.

It allows instantiating a `Daemon` from a `DeamonConfig` in the case of a standalone one, thus creating the Zenoh session based on the configuration passed in the `DaemonConfig`.
While in the case of a plugin the Zenoh sessions come from the router, so we need to use that one instead of creating a new one.